### PR TITLE
ES|QL Add initial support for semantic_text field type

### PR DESCRIPTION
**Pull request by ioanatia**
Support is added behind a feature flag. We could not simply use `EsqlCapabilities` since that's not available in `esql-core`.
Right now we have no support in existing functions.
I followed the approach for adding initial support for `date_nanos` which was also added behind a feature flag which allowed for incremental progress, rather than adding support for everything in one PR: https://github.com/elastic/elasticsearch/pull/110205

With this PR we will return `semantic_text` fields as part of the results and it will also allow us to refer to `semantic_text` fields in the `match` function (to run semantic search):

```
{
  "columns": [
    {
      "name": "semantic_text_field",
      "type": "semantic_text"
    }
  ],
  "values": [
    [
      "these are not the droids you're looking for. He's free to go around"
    ],
    [
      "all we have to decide is what to do with the time that is given to us"
    ],
    [
      "I have no idea what I'm doing, but I know I'm doing it really, really well."
    ]
  ]
}
```

realtes: #115103

### CODE REVIEW

File: x-pack/plugin/esql/qa/server/multi-clusters/src/javaRestTest/java/org/elasticsearch/xpack/esql/ccq/Clusters.java
Comment (ioanatia) [inline x-pack/plugin/esql/qa/server/multi-clusters/src/javaRestTest/java/org/elasticsearch/xpack/esql/ccq/Clusters.java:47]: this plugin is used for testing `semantic_text` - it's an inference service that can create sparse or dense embeddings - that have no actual "semantic" meaning since they are not using a model, but they are supposed to be deterministic.

File: x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java
Comment (ioanatia) [inline x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java:267]: this is the inference service from the `inference-service-test` test plugin

File: x-pack/plugin/esql/qa/server/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/multi_node/EsqlSpecIT.java
Comment (nik9000) [inline x-pack/plugin/esql/qa/server/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/multi_node/EsqlSpecIT.java:17]: Will everything need this plugin? should it be in `testCluster`?

File: x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java
Comment (nik9000) [reply x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java:267]: This one's probably worth javadoc to explain that it's for the semantic text fields.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec
Comment (nik9000) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec:4]: You'll have to `| SORT` this or else it'll be jumbled in any multi-node tests.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java
Comment (nik9000) [inline x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:580]: Someone else'll have to verify that this is the right way to load these. It's how you load text, after all. But if there's a nicer docvalues way this is stored we should use that if we can.

**Review by nik9000 — COMMENTED** [review]
Makes sense to me. @not-napoleon might want to take a look because this is beginning to follow your lead.

**Review by mark-vieira — APPROVED** [review]

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java
Comment (tteofili) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:580]: yeah, I think this is somewhat dependent on how `semantic_text` fields store text in ES/Lucene. 

In a ideal world it would be nice if this block loader doesn't have to be changed if/when `semantic_text` fields store text in different way(s) and/or change suffix.

**Review by tteofili — APPROVED** [review]

File: x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java
Comment (ChrisHegarty) [reply x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java:267]: It kinda surprises me that this is the first capability conditional in the test infra.  How expensive is this service endpoint? Conditionally creating it is of course fine, but the test infra would be simpler if it was unconditionally registered.

**Review by ChrisHegarty — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec
Comment (ChrisHegarty) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec]: good catch. done.

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java
Comment (carlosdelest) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:580]: Text in `semantic_text` is not stored. It has to be retrieved from source.

File: x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java
Comment (carlosdelest) [reply x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java:267]: The test endpoint is quite lightweight. We could always register it.

**Review by carlosdelest — APPROVED** [review]
`semantic_text` support, yay! 💯

File: x-pack/plugin/esql/qa/server/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/multi_node/EsqlSpecIT.java
Comment (ioanatia) [reply x-pack/plugin/esql/qa/server/multi-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/multi_node/EsqlSpecIT.java:17]: I think only this one needs it - it's the only one for `multi_node` that extends from `EsqlSpecTestCase` that loads the CSV data sets.

**Comment by ioanatia** [comment]
There are some issues with loading this plugin for multi-cluster and mixed-versions when one of the cluster nodes in on 8.16.0:
```
       Caused by:
        java.lang.RuntimeException: Execution of elasticsearch-plugin failed with exit code 70
            at org.elasticsearch.test.cluster.local.AbstractLocalClusterFactory$Node.runToolScript(AbstractLocalClusterFactory.java:795)
            at org.elasticsearch.test.cluster.local.AbstractLocalClusterFactory$Node.installPlugins(AbstractLocalClusterFactory.java:633)
            at org.elasticsearch.test.cluster.local.AbstractLocalClusterFactory$Node.start(AbstractLocalClusterFactory.java:159)
            at org.elasticsearch.test.cluster.local.DefaultLocalClusterHandle.lambda$start$0(DefaultLocalClusterHandle.java:78)
            at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
            at java.base/java.util.AbstractList$RandomAccessSpliterator.forEachRemaining(AbstractList.java:722)
            at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:556)
            at java.base/java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:291)
            at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:759)
            at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
            at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
            at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)
            at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035)
            at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
```

multi-node and single-node should continue to work just fine.
For now I conditionally disabled loading the plugin in multi-cluster and mixed-versions when one of the cluster nodes is on 8.x. Once we get a fix for the plugin, we can remove this check, but I don't have that much familiarity with the plugin and don't know how to fix it. I reached out to the team that owns it.

The checks `CsvTestsDataLoader` have also been changed - they no longer check for `SEMANTIC_TEXT_TYPE` ESQL capability, but rather check whether a dataset requires the test inference endpoint. This kind of makes more sense than to check for an ES|QL capability, we are not using ES|QL APIs to load the data, but the create index and bulk APIs.
If the test inference endpoint does not exist and a dataset requires it, we simply do not load the dataset.
In `EsqlSpecTestCase` I added another check for `shouldSkipTest` - in case `SEMANTIC_TEXT_TYPE` is required we also check that we can use the test inference service. This makes sense because if the test inference service we have no data with `semantic_text` fields.

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

File: x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java
Comment (ChrisHegarty) [reply x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/mapper/SemanticTextFieldMapper.java:599]: 👍

File: x-pack/plugin/esql/qa/server/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/Clusters.java
Comment (ChrisHegarty) [inline x-pack/plugin/esql/qa/server/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/Clusters.java:30]: For my one knowledge, why can the service not be loaded on ES 8.x ? Is this a short-term limitation or something else?

Comment (ioanatia) [reply x-pack/plugin/esql/qa/server/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/Clusters.java:30]: It is a short-term limitation until we fix the plugin

Comment (ioanatia) [reply x-pack/plugin/esql/qa/server/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/Clusters.java:30]: the only consequence for now is that we won't execute the test from `x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec` for mixed-version and multi-cluster

we do execute it for multi-node and single-node.
This only seems to happen when we try to set the plugin for clusters with 8.x nodes.

Comment (ChrisHegarty) [reply x-pack/plugin/esql/qa/server/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/Clusters.java:30]: Ok, got it. This workaround for the temporary limitation seems reasonable. Is there an issue tracking the plugin issue?    

Also, just to confirm. Does the test plugin work ok on an 8.x only cluster - not mixed, just 8.x nodes. For example, if this change were to be backported to the Elasticsearch 8.x branch.

**Comment by ioanatia** [comment]
@elasticmachine update branch

**Comment by ioanatia** [comment]
@elasticmachine test this please

**Comment by ioanatia** [comment]
@elasticmachine update branch

File: x-pack/plugin/esql/qa/server/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/Clusters.java
Comment (ioanatia) [reply x-pack/plugin/esql/qa/server/mixed-cluster/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/mixed/Clusters.java]: We have another plugin `test-infererence-plugin` which is used to test retrievers for which we don't have any mixed version clusters or multi cluster tests.
Because of this I think it is fine we figure out how to enable these tests as a follow up and potentially find a solution we can use not just for ES|QL integration tests.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/mapping-semantic_text.json
Comment (fang-xing-esql) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/mapping-semantic_text.json:1]: I wonder if we can add a few more fields into the mapping, like `KEYWORD`, `TEXT`, some numeric types etc? The mapping looks a bit simple to me. If `SEMANTIC_TEXT` will be in the same category as `KEYWORD` and `TEXT`, we might want to support some functions that apply to them as well(perhaps in future PRs). It will be handy to have some extra fields in the same mapping to validate how  processing commands and functions work with the new data type, like eval/where/keep/drop/sort/stats/rename/mv_expand etc.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv
Comment (fang-xing-esql) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv:1]: I wonder if `SEMANTIC_TEXT` makes sense for multi-valued field? It might be nice to have some multi-valued data to experiment with.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec
Comment (fang-xing-esql) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec:1]: Can we add a few more queries with more processing commands here, like eval/where/keep/drop/sort/stats/rename/mv_expand etc? It seems like some of the processing commands might already work out of the box, and having more tests to show the coverage will help to clarify the scope of the PR. #110205 enabled date_nano on a few scalar functions, it seems like we are not there yet, but perhaps it works in a few processing commands already, it will be great if the tests can show more besides `SORT`.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ResponseValueUtils.java
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ResponseValueUtils.java:117]: Minor suggestion, swap `SEMANTIC_TEXT` and `TEXT`.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/PositionToXContent.java
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/PositionToXContent.java:92]: Minor suggestion, swap `SEMANTIC_TEXT` and `TEXT`.

**Review by fang-xing-esql — COMMENTED** [review]
I left the first round of comments, they are mainly about the scope of this PR. Adding a new data type in ES|QL is not trivial, it makes sense to break it down to multiple steps. If I understand it correctly(please correct me if I'm wrong), this PR enables `SEMANTIC_TEXT` in **some** processing commands, if there are more tests to show the coverage and scope that will be very appreciated.

I'll need more time to look into inference test service related code.

Thank you for labelling `test-release`, it takes a little longer to run, however it did help me catch some unexpected errors.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ResponseValueUtils.java
Comment (ioanatia) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ResponseValueUtils.java:117]: No problem with changing this - but I am curious what's the reason? Right now they are in alphabetical order.
If someone takes a look at this code a year from now, they might appreciate having these in alphabetical order rather than having `SEMANTIC_TEXT` at the end which specifies that support was added for `SEMANTIC_TEXT` after `KEYWORD` and `TEXT` are available.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec
Comment (ioanatia) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv-spec:1]: added tests for `eval/where/keep/drop/sort/stats/rename/mv_expand`
for functions and operators there is no support - I am doing that as a follow up here https://github.com/ioanatia/elasticsearch/compare/semantic_text_support...ioanatia:elasticsearch:semantic_text_as_string?expand=1
I took the dataset/mapping I had in my follow-up branch and added it here - the dataset has many fields that are not used in the tests yet, but will be used to test things like conversion functions when we add support for those.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv
Comment (ioanatia) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/semantic_text.csv]: Yes it absolutely does! I was not sure whether I should add a test for it here, but I did anyway for `MV_EXPAND`

**Comment by ioanatia** [comment]
@elasticmachine update branch

File: x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java
Comment (ioanatia) [reply x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java:266]: javadoc was added

**Review by fang-xing-esql — COMMENTED** [review]
I left a few more comments. This is a subtask of supporting `SEMANTIC_TEXT` in `ES|QL`. I wonder if there is an umbrella issue created to keep track of all of the subtasks, similar to `DATE_NANOS`, so that we know where we are, and what to expect?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ResponseValueUtils.java
Comment (ChrisHegarty) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/ResponseValueUtils.java:117]: When I wrote this code originally, and many other similar switches in the compute engine, I deliberately did not follow a strict style consistency across the codebase, but rather preferred a logical and type grouping style. With a preference for more common and narrower types earlier in the cases.
I have no strong preference of ordering of this particular case label here.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/PositionToXContent.java
Comment (ChrisHegarty) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/PositionToXContent.java:92]: see https://github.com/elastic/elasticsearch/pull/113920#discussion_r1806171017

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_unsupported_types.yml
Comment (ChrisHegarty) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_unsupported_types.yml:528]: The compute engine is not actually doing any inference calls at this point, so the value here is for the most part irrelevant. However, it should likely be the same as that of what was configured by in the inference service (see `createInferenceEndpoint`), so `test_sparse_inference`.

**Comment by ChrisHegarty** [comment]
> I wonder if there is an umbrella issue created to keep track of all of the subtasks, similar to `DATE_NANOS`, so that we know where we are, and what to expect?

Good idea. Done #115103

**Comment by ChrisHegarty** [comment]
FYI - I am removing the `test-release` label for now. It's adding too much latency and slowing iteration on this PR. It will be added before merge, at which time any failing tests it catches will be fixed.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/mapping-semantic_text.json
Comment (ChrisHegarty) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/mapping-semantic_text.json:1]: done [e617196](https://github.com/elastic/elasticsearch/pull/113920/commits/e617196bfc36c52a17b3eb8bbc28c193924a4975).  More field types and functions can come later.

File: x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java
Comment (ChrisHegarty) [reply x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/EsqlSpecTestCase.java]: I fixed the registration of the inference plugin for mixed and multi-cluster tested.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_unsupported_types.yml
Comment (ChrisHegarty) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_unsupported_types.yml:528]: There is no lookup of inference yet. That will come in a subsequent step (see the meta issue).

**Comment by ChrisHegarty** [comment]
Ok, mixed node testing will be avoided until the following issue is resolved as a follow up. #115166.  (given that mixed mode will only be relevant when this PR is backported)

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Review by ChrisHegarty — APPROVED** [review]
This PR has had a number of review iterations. It's a great first step towards support for semantic search in ES|QL. LGTM 👍

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_unsupported_types.yml
Comment (ChrisHegarty) [reply x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/40_unsupported_types.yml:543]: This is just how one helps the test infra translate to embeddings. It's mostly not interesting until the engine actually attempts the inference. Which will come in a later stage. (see the meta issue).

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ChrisHegarty** [comment]
@elasticmachine update branch

**Comment by ioanatia** [comment]
@elasticmachine test this please

**Comment by ioanatia** [comment]
@elasticmachine update branch

**Comment by ioanatia** [comment]
Thank you folks, especially @fang-xing-esql and @ChrisHegarty for your help on this PR!
All the comment have been addressed, CI is green, the PR has a final approval from Chris after we addressed the comments, so it looks good to go 🚀

**Pull request merged**
